docs/tutorials/10_torch_qgan.ipynb

```diff
@@ -0,0 +1,752 @@
+    }
+   },
+   "source": [
+    "# PyTorch qGAN Implementation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "Description\n",
+    "\n",
+    "adapted from [PyTorch GAN](https://github.com/eriklindernoren/PyTorch-GAN/blob/master/implementations/gan/gan.py)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 43,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Necessary imports\n",
+    "\n",
+    "import numpy as np\n",
+    "from typing import Union, List, Optional, Iterable\n",
+    "\n",
+    "from torch import Tensor, stack, reshape\n",
+    "from torch.utils.data import DataLoader\n",
+    "from torch.autograd import Variable\n",
+    "from torch.optim import Adam\n",
+    "import torch.nn as nn\n",
+    "import torch.nn.functional as F\n",
+    "\n",
+    "\n",
+    "from qiskit import Aer, QuantumCircuit\n",
+    "from qiskit.utils import QuantumInstance, algorithm_globals\n",
+    "from qiskit.opflow import Gradient, StateFn\n",
+    "from qiskit.circuit.library import TwoLocal\n",
+    "from qiskit.circuit import ParameterExpression, ParameterVector\n",
+    "from qiskit_machine_learning.neural_networks import CircuitQNN\n",
+    "from qiskit_machine_learning.connectors import TorchConnector\n",
+    "from qiskit_machine_learning.datasets.dataset_helper import discretize_and_truncate\n",
+    "\n",
+    "# Set seed for random generators\n",
+    "algorithm_globals.random_seed = 42\n",
+    "np.random.seed = 42"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "### Load training data\n",
+    "\n",
+    "For testing purposes, we decide for a 2D multivariate normal distribution.\n",
+    "Each dimension is represented by 2 qubits."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 44,
+   "outputs": [],
+   "source": [
+    "data_dim = [3, 3]\n",
+    "\n",
+    "training_data = np.random.default_rng().multivariate_normal(mean=[0., 0.], cov=[[1, 0], [0, 1]], size=1000, check_valid='warn',\n",
+    "                                                        tol=1e-8, method='svd')\n",
+    "# Define minimal and maximal values for the training data\n",
+    "bounds_min = np.percentile(training_data, 5, axis=0)\n",
+    "bounds_max = np.percentile(training_data, 95, axis=0)\n",
+    "bounds = []\n",
+    "for i, _ in enumerate(bounds_min):\n",
+    "    bounds.append([bounds_min[i], bounds_max[i]])\n",
+    "\n",
+    "# Pre-processing, i.e., gridding\n",
+    "(training_data,\n",
+    "data_grid,\n",
+    "grid_elements,\n",
+    "prob_data ) = discretize_and_truncate(\n",
+    "training_data,\n",
+    "np.array(bounds),\n",
+    "data_dim,\n",
+    "return_data_grid_elements=True,\n",
+    "return_prob=True,\n",
+    "prob_non_zero=True,\n",
+    ")\n",
```

Comment:
this section can be simplified if you don't return all of these extra elements you will not use (grid, etc). Also, if you apply the suggestion above about bounds, it's already a bumpy array, so no need to cast:

```suggestion
    "training_data, _ = discretize_and_truncate(data = training_data, \n"
     "                                     min_max_bin_centers = bounds, \n"
     "                                      num_qubits = data_dim), \n"
```